Enhance docs with GitBook

dpordomingo opened the pull request:
fix #330
blocks #423 and #424

The GitBook docs site is available under my own fork:
https://dpordomingo.gitbook.io/lookout

Once it is approved, it would be made public under:
https://docs.sourced.tech/lookout

I did:
- created the GitBook structure and complete the docs according to the team agreements,
- added the `.github/CODE_OF_CONDUCT.md` file,
- added the `DCO` file,
- fixed the `LICENSE.md` file, and created the `NOTICE.md` file,
- moved the `CONTRIBUTING.md` file into `docs` and enhance it,
- created a `docker-compose-rabbitmq.yml` so RabbitMQ dependency will be runnable as the others,
- move the the sequence diagrams into `docs/assets/` and enhance them,
- avoid downloading the toc generator if already available,

dpordomingo commented:
I rebased on top of #394 to fix Travis failing tests :dagger:

dpordomingo commented:
I added two new commits:
- 00bd5a8 Explain how to stop lookout and restore the DB
- fdd5d73 Explain the most important things to be configured

carlosms commented on docs/analyzers-creation.md:
```diff
@@ -0,0 +1,74 @@
+
+## Fetching Changes, UASTs or Languages from lookout DataService
+
+The analyzer should not need to take care of managing git data, nor storing the repositories, nor guessing the programming language used in the files nor parsing its UASTs; to do so, it should request **lookout DataService** for such data.
```
It sounds a bit convoluted to me. What about
```suggestion
Lookout will take care of dealing with Git repositories, UAST extraction, programming language detection, etc. Your analyzer will be able to use the **lookout DataService** to query all this data.
```

carlosms commented on docs/analyzers-creation.md:
```diff
@@ -0,0 +1,74 @@
+
+The analyzer should not need to take care of managing git data, nor storing the repositories, nor guessing the programming language used in the files nor parsing its UASTs; to do so, it should request **lookout DataService** for such data.
+
+You can read more about it in [**lookout DataService** section](architecture.md#lookout-dataservice).
```
```suggestion
You can read more about it in the [**lookout DataService** section](architecture.md#lookout-dataservice).
```

carlosms commented on docs/analyzers-creation.md:
```diff
@@ -0,0 +1,74 @@
+You can read more about it in [**lookout DataService** section](architecture.md#lookout-dataservice).
+
+
+## How to Locally Try an Analyzer
```
```suggestion
## How to Test an Analyzer Locally
```

carlosms commented on docs/analyzers-creation.md:
```diff
@@ -0,0 +1,74 @@
+
+# Using Pregenerated Code from the SDK
+
+If you're creating the analyzer in Golang or Python, you'll find pre-generated libraries in the [lookout-sdk repository](https://github.com/src-d/lookout-sdk) so you'll have to generate nothing.
```
I would add something about the helpers. What about:

```suggestion
If you're creating your analyzer in Golang or Python, you'll find pre-generated libraries in the [lookout-sdk repository](https://github.com/src-d/lookout-sdk). The SDK libraries also come with helpers to deal with gRPC caveats.
```

carlosms commented on docs/analyzers-creation.md:
````diff
@@ -0,0 +1,74 @@
+}
+```
+
+Once the gRPC server is run, it will listen for requests from `lookoutd`.
````
```suggestion
Once the gRPC server is started, it will listen for requests from `lookoutd`.
```

carlosms commented on docs/analyzers-creation.md:
```diff
@@ -0,0 +1,74 @@
+
+## Golang
+
+Import and use `github.com/src-d/lookout-sdk/pb`.
```
```suggestion
Import and use `gopkg.in/src-d/lookout-sdk.v0/pb`.
```

carlosms commented on docs/analyzers-creation.md:
```diff
@@ -0,0 +1,74 @@
+
+Import and use `github.com/src-d/lookout-sdk/pb`.
+
+The analyzer must implement the [AnalyzerClient interface](https://github.com/src-d/lookout-sdk/blob/master/pb/service_analyzer.pb.go#L129):
```
This should be AnalyzerServer, not the client, right?
Also, maybe it's better to link to godocs.

```suggestion
The analyzer must implement the [AnalyzerServer interface](https://godoc.org/gopkg.in/src-d/lookout-sdk.v0/pb#AnalyzerServer):
```

carlosms commented on docs/analyzers-creation.md:
````diff
@@ -0,0 +1,74 @@
+
+The analyzer must implement the [AnalyzerClient interface](https://github.com/src-d/lookout-sdk/blob/master/pb/service_analyzer.pb.go#L129):
+
+```go
````
I propose this other code:

```go
import "gopkg.in/src-d/lookout-sdk.v0/pb"


type Analyzer struct {}

func (a *Analyzer) NotifyReviewEvent(ctx context.Context, e *lookout.ReviewEvent) (
...
}

func (a *Analyzer) NotifyPushEvent(ctx context.Context, e *lookout.PushEvent) (*lookout.EventResponse, error) {
...
}

smacker replied:
👍 current code is invalid

carlosms replied:
proposed*

And yes, LGTM :+1:

carlosms commented on docs/analyzers-creation.md:
````diff
@@ -0,0 +1,74 @@
+$ pip install lookout-sdk
+```
+
+The analyzer class will extend [AnalyzerServicer](https://github.com/src-d/lookout-sdk/blob/master/python/lookout/sdk/service_analyzer_pb2_grpc.py#34):
````
Bad link,
```suggestion
The analyzer class will extend [AnalyzerServicer](https://github.com/src-d/lookout-sdk/blob/master/python/lookout/sdk/service_analyzer_pb2_grpc.py#L34):
```

smacker replied:
in general it's bad idea to link on line on master. It will become wrong sooner or later. Better link to the last tag. And just grep when we release new version.

carlosms commented on docs/analyzers-examples.md:
```diff
@@ -0,0 +1,23 @@
+
+_Please refer to [**lookout-sdk** docs](lookout-sdk.md) to see how to locally test an analyzer without accessing GitHub._
+
+**lookout-sdk** repository contains a [quickstart example](https://github.com/src-d/lookout-sdk/blob/master/examples) &mdash;implemented in Go and in Python&mdash; of an Analyzer that detects the language and number of functions for every file.
```
```suggestion
The **lookout-sdk** repository contains a [quickstart example](https://github.com/src-d/lookout-sdk/blob/master/examples) &mdash;implemented in Go and in Python&mdash; of an Analyzer that detects the language and number of functions for every file.
```

carlosms commented on docs/analyzers-examples.md:
```diff
@@ -0,0 +1,23 @@
+
+## Golang
+
+You can run [language-analyzer.go](https://github.com/src-d/lookout-sdk/blob/master/examples/language-analyzer.go) running from **lookout-sdk** directory:
```
```suggestion
You can execute [language-analyzer.go](https://github.com/src-d/lookout-sdk/blob/master/examples/language-analyzer.go) running from the **lookout-sdk** directory:
```

carlosms commented on docs/analyzers-examples.md:
```diff
@@ -0,0 +1,23 @@
+
+## Python
+
+You can run [language-analyzer.py](https://github.com/src-d/lookout-sdk/blob/master/examples/language-analyzer.py) running from **lookout-sdk** directory:
```
```suggestion
You can execute [language-analyzer.py](https://github.com/src-d/lookout-sdk/blob/master/examples/language-analyzer.py) running from the **lookout-sdk** directory:
```

smacker commented on README.md:
```diff
@@ -1,209 +1,165 @@
+**source{d} Lookout** is the first step towards a full suite of Machine Learning on Code applications for AI-assisted coding, but you can also create your own analyzers without an ML approach.
 
-Please refer to the [**Running in Distributed Mode guide**](docs/distributed.md) for documentation on the advanced distributed deployment.
+The benefits of using **source{d} lookout** are:
```
not sure if it's still true but before we had an agreement to use lower case for all source{d} names including source{d}.

smacker commented on README.md:
````diff
@@ -1,209 +1,165 @@
 ```
 
-## Code generation
+You can stop it pressing `ctrl+C`
````
inconsistency: `ctrl+C` here and `ctrl+c` on the next line

smacker commented on docs/CONTRIBUTING.md:
````diff
@@ -0,0 +1,99 @@
+
+```bash
+$ kallax migrate --input ./store/models/ --out ./store/migrations --name <name>
+$ make dependencies
````
wrong order of commands. Without `make dependencies` there is no kallax in $GOPATH/bin.

smacker commented on docs/analyzers-creation.md:
````diff
@@ -0,0 +1,74 @@
+The analyzer class will extend [AnalyzerServicer](https://github.com/src-d/lookout-sdk/blob/master/python/lookout/sdk/service_analyzer_pb2_grpc.py#34):
+
+```python
+class AnalyzerServicer(object):
````
if we change go example above we need to change python too

smacker commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+
+In some situations, you might want to run all its dependencies and components separately, or you can even want to [run it in a distributed mode](#running-lookout-in-distributed-mode). Both ways to run lookout would require to follow these general steps:
+
+1. [Run lookout dependencies](#dependencies),
```
similar to the comments before, let's drop `,`

smacker commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+## Dry-run Mode
+If you want to avoid posting the analysis results on GitHub, and only print them, enable the _dry-run_ mode when running `serve`, `work` subcommands:
+
+| subcommands | Env var | Option |
```
looks like tables are broken:
https://github.com/src-d/lookout/blob/fdd5d732139120ce7e87f2e370de9ec6c55fbd0f/docs/how-to-run.md

dpordomingo replied:
whaaat? how did I ruin them!!?
thanks for spotting that problem!

dpordomingo replied:
fixed by 3912bed

dpordomingo replied:
fixed by e3d47eb

carlosms commented on docs/architecture.md:
```diff
@@ -0,0 +1,46 @@
+
+_You will find the project background, goals, constraints, solution details and other info in [**source{d} lookout DesignDoc**](https://docs.google.com/document/d/1pqz-_SHO5BsJE-aa8o_bAY3r5vR67amnWN8-qZc2UgY/edit#)_
+
+![lookout service sequence diagram](assets/lookout-seq-diagram.png)
```
I don't like this diagram at all, to be honest. This one was in the docs at some point, then we removed it. Well, we moved it to a `.md` file that was not linked from anywhere, but should have been deleted.

I propose to replace it with this other one: [link](https://mermaidjs.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG4gICAgcGFydGljaXBhbnQgR2l0SHViXG4gICAgcGFydGljaXBhbnQgbG9va291dFxuICAgIHBhcnRpY2lwYW50IEFuYWx5emVyXG4gICAgcGFydGljaXBhbnQgQmFiZWxmaXNoXG4gICAgbG9va291dC0-PkdpdEh1YjogUG9sbGluZ1xuICAgIEdpdEh1Yi0tPj5sb29rb3V0OiBQUi9QdXNoIGV2ZW50c1xuICAgIGxvb2tvdXQtPj5BbmFseXplcjogTm90aWZ5UmV2aWV3RXZlbnRcbiAgICBBbmFseXplci0-Pmxvb2tvdXQ6IEdldENoYW5nZXMvR2V0RmlsZXNcbiAgICBsb29rb3V0LT4-QmFiZWxmaXNoOiBwYXJzZUZpbGUgKGlmIFdhbnRVQVNUKVxuICAgIEJhYmVsZmlzaC0tPj5sb29rb3V0OiBVQVNUXG4gICAgbG9va291dC0tPj5BbmFseXplcjogQ2hhbmdlL0ZpbGVcbiAgICBBbmFseXplci0tPj5sb29rb3V0OiBDb21tZW50c1xuICAgIGxvb2tvdXQtPj5HaXRIdWI6IFBvc3QgY29tbWVudHMiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9fQ).

In my opinion, this one is overly-complicated. The goal of this diagram should be to provide a simple overview of the different components and how they interact with each other.
- What is that healthcheck loop exactly? And how is that relevant for an overview of the architecture?
- The registered repos/registered analyzers boxes add noise, without any added value
- DataServer is shown as a component. While this might be useful to describe the internals of the `lookoutd` code, from an architecture point of view this adds confusion. `Analyzer` could also be divided into `Analyzer gRPC Server`, `Analyzer` and `DataServer gRPC client`; but that would be just unnecesary noise. I think separating `lookout` and `DataServer` is in a similar category, too low-level for a diagram.

carlosms replied:
About
> boxes are loop according to the label they have, meaning that something happens for every repo and every analyzer... because those events are not being processed by different lookout instances, but for the same one.
 
The boxes are correct, and I don't find them not misleading in any way. My feedback here is that visually it makes the diagram too busy, making it harder to understand things at a glance.

In this diagram I would say less is more. Saying that we poll GitHub for the registered repos is implied, just like we don't need to say that we'll use the configured credentials to poll GitHub. Or that we'll do that periodically in a loop; I think that's also imiplied.

dpordomingo replied:
done by a178a5028745b566f4b013e29b668649805b754a

carlosms commented on docs/architecture.md:
```diff
@@ -0,0 +1,46 @@
+It takes review requests made by an external code review system, calls the registered analyzers to review the changes, and posts the results back.
+
+
+## lookout DataService
```
For the same reasoning as above, I would not consider `DataService` a component. I would simply explain this as one of the things that the lookout server does.

This is a personal opinion on what is a _component_. But from an architecture point of view, I would stop at `lookout`.

For someone that deploys lookout, an analyzer needs to call the `lookoutd` component, using the `DataService` gRPC service definition. For the same reason we have the `lookout` component contact GitHub, and not a separate `github provider`, or `go-github` component.

dpordomingo replied:
I also found that "DataService" is part of the API that the user is using, so it's may worth to explain who's attending their requests and how is that related to Lookout.
In a further iteration, DataService could be even physically separated, and this info would be still meaningful.

dpordomingo replied:
addressed by 3484af7

carlosms commented on docs/architecture.md:
```diff
@@ -0,0 +1,46 @@
+
+They are not part of **lookout** repository so they can be developed by third parties.
+
+Lookout Server will call all the Analyzers that were already registered when it was started.
```
I don't know why exactly but this sentence sounds a bit off. What about
```suggestion
Lookout Server will call all the Analyzers to produce comments for a new Pull Request. To register new Analyzers in the configuration file, lookout will need to be restarted.
```

carlosms commented on docs/architecture.md:
```diff
@@ -0,0 +1,46 @@
+
+The [lookout-sdk](https://github.com/src-d/lookout-sdk) repository is a toolkit for writing new analyzers. It contains:
+- `.proto` interface definitions for all **lookout** gRPC services.
+- The pre-generated code in Go and Python that provides an easy access to the **lookout DataService** gRPC service; and low-level helpers to workaround some protobuf/gRPC caveats.
```
```suggestion
- The pre-generated code in Go and Python that provides an easy access to the **lookout DataService** gRPC service; and low-level helpers to work around some protobuf/gRPC caveats.
```

smacker commented:
The introduction looks strange on gitbook. On my laptop unnecessary blank space, logo, and so on take the whole first screen:
<img width="1552" alt="screenshot 2018-12-11 at 17 28 07" src="https://user-images.githubusercontent.com/406916/49814559-2857bf00-fd6a-11e8-8139-f9fa453c6de2.png">

When I compare it with [gitbase](https://docs.sourced.tech/gitbase) or [bblfsh](https://docs.sourced.tech/babelfish) it looks MUCH worse.

smacker commented on docs/analyzers.md:
```diff
@@ -0,0 +1,34 @@
+# source{d} lookout Analyzers
```
look like the name on gitbook is different. And I think it reflects the content of this file better.

smacker commented on docs/configuration.md:
```diff
@@ -76,7 +89,9 @@ The minimum watch interval to discover new pull requests and push events is defi
-The list of repositories to be watched by **lookout** is defined by the `repositories` key.
+The list of repositories to be watched by **lookout** is defined by:
+- the `repositories` field at `config.yml`, or
+- automatically the GitHub App installation if you [authenticated as a GitHub App](#github-app); in that case, the `repositories` field at `config.yml` will be ignored.
```
```
automatically the GitHub App installation if
```
sounds wrong

carlosms replied:
```suggestion
- the repositories where the GitHub App is installed if you [authenticated as a GitHub App](#github-app). In that case, the `repositories` field in `config.yml` will be ignored.
```

smacker commented on docs/analyzers-creation.md:
```diff
@@ -0,0 +1,74 @@
+# Implementing Your Own Analyzer
```
I wish we had links to examples somewhere in this document. Because after reading it I still have no idea how to start.

dpordomingo replied:
Was it improved with the provided examples?
eb222fb

smacker replied:
yes, definitely! Though python example is incorrect (it exits after 60 seconds).

dpordomingo replied:
I think it's up to the user to decide how to proceed to wait before stopping the server.

I also tried to be concise, so I preferred not to handle that in detail.
On the other hand, it is also explained in the gRPC.io docs about how to start the server, already linked in our docs just above the python example.

smacker commented:
Overall 👍

carlosms commented on docs/configuration.md:
```diff
@@ -12,12 +12,25 @@ repositories:
 analyzers:
     # list of named analyzers
+timeout:
+    # configuration for the existent timeouts.
```
```suggestion
    # configuration for the existing timeouts.
```

carlosms commented on docs/configuration.md:
```diff
@@ -12,12 +12,25 @@ repositories:
+The most important things you need to configure for a local installation, are:
+
+1. [Repositories](#repositories): add the URLs of the repositories to be watched.
+1. [Analyzers](#analyzers): add the gRPC addresses of the analyzers to be used by lookout,
```
```suggestion
1. [Analyzers](#analyzers): add the gRPC addresses of the analyzers to be used by lookout.
```

carlosms commented on docs/configuration.md:
```diff
@@ -36,17 +49,17 @@ providers:
+**lookout** needs to authenticate with GitHub. The easiest method for testing purposes is using a [GitHub personal access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/). The token should have the `repo` scope enabled.
 
-The credential can then be passed to `lookoutd` as:
+The credential can then be also passed to `lookoutd` as:
```
Why did you add `also`? In addition to what?

carlosms replied:
I see. But if you read this sentence, it says "also" when there is no previous alternative. This is the first one, it sounds weird to say "also".

dpordomingo replied:
I think now is more clear: 6ff7b2411587c8f551326c3d30107be0b6573315
Any other idea @carlosms

carlosms commented on docs/configuration.md:
```diff
@@ -146,23 +159,25 @@ The `settings` for each analyzer in the `.lookout.yml` config file will be merge
 # These are the default timeout values. A value of 0 means no timeout
 timeout:
-  # Timeout for an analyzer response to a NotifyReviewEvent
+  # Timeout for an analyzer to response a NotifyReviewEvent
```
I'd leave "Timeout for an analyzer response to a NotifyReviewEvent". Or change this to "to reply".

carlosms commented on docs/configuration.md:
```diff
@@ -146,23 +159,25 @@ The `settings` for each analyzer in the `.lookout.yml` config file will be merge
+  # Timeout for an analyzer to response a NotifyPushEvent
   analyzer_push: 60m
-  # Timeout http requests to the GitHub API
+  # Timeout for an HTTP requests to the GitHub API
```
```suggestion
  # Timeout for HTTP requests to the GitHub API
```

carlosms commented on docs/configuration.md:
```diff
@@ -146,23 +159,25 @@ The `settings` for each analyzer in the `.lookout.yml` config file will be merge
   # Timeout for Git fetch actions
   git_fetch: 20m
-  # Timeout for parse requests to Bblfsh
+  # Timeout for Bblfsh to response a Parse request
```
```suggestion
  # Timeout for Bblfsh to reply to a Parse request
```

carlosms commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+
+The recommended way to locally try **source{d} lookout** is using `docker-compose` as described in [quickstart documentation](../README.md#quickstart).
+
+In some situations, you might want to run all its dependencies and components separately, or you can even want to [run it in a distributed mode](#running-lookout-in-distributed-mode). Both ways to run lookout would require to follow these general steps:
```
```suggestion
In some situations, you might want to run all its dependencies and components separately, or you may even want to [run it in a distributed mode](#running-lookout-in-distributed-mode). Both ways to run lookout would require to follow these general steps:
```

carlosms commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+1. [Configure lookout](#configure-lookout),
+1. [Run lookout](#run-lookout).
+
+In other situations you just want to locally test an analyzer without accessing GitHub at all; for that situations, you might want to read [`lookout-sdk` binary documentation](lookout-sdk.md).
```
```suggestion
In other situations you may just want to test an analyzer locally without accessing GitHub at all. For those situations, you might want to read the [`lookout-sdk` binary documentation](lookout-sdk.md).
```

carlosms commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+lookout can be also run in a distributed fashion using a [RabbitMQ](https://www.rabbitmq.com/) queue to coordinate a _watcher_ and several _workers_.
+
+- The _watcher_ process monitors GitHub pull requests and enqueues jobs for new events.
+- The running _workers_ dequeue jobs as they become available calling the registered analyzers and posting its returned comments.
```
```suggestion
- The running _workers_ dequeue jobs as they become available, call the registered analyzers, and post the results as comments.
```

carlosms commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+**lookout** needs a running instance of:
+
+* [bblfshd](https://github.com/bblfsh/bblfshd) to parse files into [UAST](https://doc.bblf.sh/uast/uast-specification.html).
+* [PostgreSQL](https://www.postgresql.org) to internally store `PushEvent`, `ReviewEvent`, `ReviewTarget` and `Comment`.
```
The list of things seems out of place. The DB is managed by lookout, as a user you don't need to know or care about `PushEvent`, `ReviewEvent`, `ReviewTarget` and `Comment`. Adding them here makes it look like it's relevant for you to know about these things, but we don't explain what is each one (because you don't really need to know about them).

In this case less is more in my opinion.
```suggestion
* [PostgreSQL](https://www.postgresql.org) for persistence.
```

carlosms commented on docs/how-to-run.md:
````diff
@@ -0,0 +1,168 @@
+In case you want to run it in a distributed way, you will also need RabbitMQ, so you can run instead:
+
+```bash
+$ docker-compose -f docker-compose.yaml -f docker-compose-rabbitmq.yml up -d --force-recreate bblfsh postgres rabbitmq
````
Do we really want to provide a docker-compose file with rabbitmq? I see its value for us, as developers. But for someone looking to deploy lookout in a distributed way, will they care about running the distributed scenario in their own laptop?

On the other hand taking a look at the compose file can be helpful to understand how to deploy each component manually. :thinking:

dpordomingo replied:
I think it worth having it in a separated file, that can be used if needed, and is consistent with how different dependencies are run.

carlosms commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+<a id=options></a>
+# Appendix: `lookoutd` Command Options
+
+`lookoutd` binary include some subcommands as described above, and they accept many different options; you can use:
```
```suggestion
`lookoutd` binary includes some subcommands as described above, and they accept many different options; you can use:
```

carlosms commented on docs/how-to-run.md:
```diff
@@ -0,0 +1,168 @@
+
+Here are some of the most relevant options for `lookoutd`:
+
+- [dry-run mode](#dry-run-mode),
```
I'd remove the trailing `,`

carlosms commented on docs/lookout-sdk.md:
```diff
@@ -0,0 +1,119 @@
+
+`lookout-sdk` binary is a simplified version of the lookout server that works with a local git repository and does not need access to Github.
+
+You can think about `lookout-sdk` as a _curl-like_ tool to call an analyzer gRPC endpoints with a ReviewEvent or a PushEvent, from a local git repository, and send it to an analyzer without accessing GitHub at all. For convenience, `lookout-sdk` also exposes a **lookout DataService** backed by the same git repository.
```
```suggestion
You can think about `lookout-sdk` as a _curl-like_ tool to call an analyzer gRPC endpoint with a `ReviewEvent` or a `PushEvent`, from a local git repository, and send it to an analyzer without accessing GitHub at all. For convenience, `lookout-sdk` also exposes a **lookout DataService** backed by the same git repository.
```

carlosms commented on docs/lookout-sdk.md:
```diff
@@ -0,0 +1,119 @@
+
+This is the sequence diagram of the `ReviewEvent` made by `lookout-sdk review`. You can compare it with a regular **lookout** installation at the [**source{d} lookout architecture** documentation](architecture.md).
+
+![sequence diagram](assets/lookout-sdk-seq-diagram.png)
```
As I commented before. I don't like that a diagram shows `DataService` as a component, when it's just an internal part of the `lookout-sdk` command. Showing it at the same level as `lookout-sdk` may lead to believe you need to run something _else_, and it's not clear what.

I would merge `DataServer` with `lookout-sdk` in this diagram.

dpordomingo replied:
addressed by 3484af7

carlosms commented on docs/lookout-sdk.md:
````diff
@@ -0,0 +1,119 @@
+
+we could run:
+
+```shell
````
Wrong command, `branch-b` is not a parent of `branch-a`

carlosms commented on docs/lookout-sdk.md:
```diff
@@ -0,0 +1,119 @@
+
+_For more options to run `lookout-sdk`, take a look into [**lookout-sdk Command Options**](#options)_
+
+- If analyzer gRPC address is omitted, it will be `ipv4://localhost:9930`,
```
I'd prefer to not use trailing `,`

carlosms commented on docs/lookout-sdk.md:
```diff
@@ -0,0 +1,119 @@
+_For more options to run `lookout-sdk`, take a look into [**lookout-sdk Command Options**](#options)_
+
+- If analyzer gRPC address is omitted, it will be `ipv4://localhost:9930`,
+- if `--git-dir` is omitted, it will be used the current dir,
```
```suggestion
- if `--git-dir` is omitted, the current dir will be used
```

carlosms commented on docs/lookout-sdk.md:
```diff
@@ -0,0 +1,119 @@
+- if `--git-dir` is omitted, it will be used the current dir,
+- if `--from` is omitted, it will be `HEAD^`,
+- if `--to` is omitted, it will be `HEAD`,
+- both `--from` and `--to` can be a tag name, branch name or the full SHA-1 object name.
```
Well... If we get picky, `HEAD` is not any of these 3, right?
What about
```suggestion
- both `--from` and `--to` can be any [git revision](https://git-scm.com/docs/gitrevisions#_specifying_revisions). For example a tag name, branch name or the full commit SHA-1.
```

carlosms commented on docs/lookout-sdk.md:
```diff
@@ -0,0 +1,119 @@
+- if `--to` is omitted, it will be `HEAD`,
+- both `--from` and `--to` can be a tag name, branch name or the full SHA-1 object name.
+
+All the explained above for `lookout-sdk review`, calling `NotifyReviewEvent`, applies also to `NotifyPushEvent` when using `lookout-sdk push`.
```
```suggestion
Everything explained above for `lookout-sdk review` calling `NotifyReviewEvent`, applies also to `NotifyPushEvent` when using `lookout-sdk push`.
```

dpordomingo commented on README.md:
```diff
@@ -1,209 +1,165 @@
+    <img src="https://godoc.org/github.com/src-d/lookout?status.svg"
+         alt="GoDoc">
+  </a>
+  <a href="https://docs.google.com/document/d/1pqz-_SHO5BsJE-aa8o_bAY3r5vR67amnWN8-qZc2UgY">
```
[source{d} Engine docs](https://docs.sourced.tech/engine) also exposes it.

carlosms replied:
Still, I think this project in its current form has surpassed the scope and solutions. My opinion is that this design doc is not ready to be used as reference material, and the current docs do a better job of describing the current architecture. Linking to it at the same level as godoc brings more confusion than benefits.

I would personally prefer not to link it from the documentation. But I'm up for having a chat about this.

dpordomingo replied:
ok, removed

smacker commented:
Feedback from @EgorBu :
Following the quick start, it's not clear what dummy analyzer does. Does it process all files? Only changed? Only parsed? What should I expect to see?

dpordomingo commented:
I addressed all your suggestions.
The ones that I didn't apply were the ones related to architecture, diagrams and `DataService`, which I'll try to reconsider tomorrow unless there is a strong strong opinion on your side ;)

dpordomingo commented:
By the way: MANY MANY THANKS FOR YOUR DETAILED REVIEWS!!
You helped me A LOT.
thank you @carlosms @smacker

dpordomingo commented:
@smacker your [concern about the header](https://github.com/src-d/lookout/pull/398#issuecomment-446267608) in GitBook was solved by 6f9eba0, and already available at my fork:
https://dpordomingo.gitbook.io/lookout

dpordomingo commented:
pending:
- [ ] fix tables from `lookout/docs/how-to-run.md`
- [ ] agree about architecture section

dpordomingo commented:
Y squashed the prev history, and added 4 new ones, addressing @carlosms requirements:
* 3484af7 Mention DataService as a gRPC service instead of as a Lookout component 
* 8bb0580 Use a simplified version of the lookout-sdk sequence diagram
* a178a50 Use a simplified version of the lookout sequence diagram

Plus:
* b57c5cb Remove SDK mention in Architecture docs
